qTox/qTox
# Move default autoaccept dir

Closes #5375

Alright this one ended up being much more complicated than I thought, handling upgrade of settings is non-trivial. That being said I think I came up with an okay solution.

On first run with the change the user sees

![new_default_dir_upgrade_global](https://user-images.githubusercontent.com/1069811/46656627-57339800-cb63-11e8-8253-04fe32f4b9a4.PNG)

*if* they previously set the autoaccept direcotry

Then for each friend that they previously enabled upgrade on they see 

![new_default_dir_upgrade_friend](https://user-images.githubusercontent.com/1069811/46656636-5e5aa600-cb63-11e8-88f0-98727c108436.PNG)

That's the major idea here, rest of the changes are in the commit message (pasted here for convenience)

> feat(transfer): Change default autoaccept directory
> 
> * Implements clean upgrades to the new defaults by requesting user
> approval to use new defaults over previously set values
> * Changes default global autoaccept directory to appDataDir()/transfers
> * Changes each default user autoaccept directory to
> appDataDir()/transfrers/hashed_friend_public_key
> * Adds buttons in the about friend dialog and general settings dialog to
> open download directory
> * Decouples friend autoaccept from autoaccept dir
>     * Interaction with global autoaccept is still broken but that will
>     be resolved in #5376 
> * Adds ability to reset friend autoaccept dir to default in the about
> friend dialog
> * Images are now saved to the per-user transfer directory instead of the global
> images folder

Screenshots of new UI elements
![new_default_dir_context_menu](https://user-images.githubusercontent.com/1069811/46656891-e93ba080-cb63-11e8-81bd-114c47357cc6.PNG)
![new_default_dir_friend_about](https://user-images.githubusercontent.com/1069811/46656892-e93ba080-cb63-11e8-8243-170c768cf59f.PNG)
![new_default_dir_settings_page](https://user-images.githubusercontent.com/1069811/46656894-e93ba080-cb63-11e8-8e31-1a2932887881.PNG)

Given that this touches the same UI elements as #5373 the UI elements may not merge cleanly, in that case just reassign whichever goes in second to me and I can update them.

Tests I ran before pushing

* Global and User auto accept set
    * Do not accept global auto accept upgrade
        * Do not accept user auto accept upgrade
            * Downloads still go to user download dir
        * Accept user auto accept upgrade
            * Downlaods go to global download dir / hashed id
                * This is reasoanble behavior since the new default user download dir is supposed to be relative to the global one
    * Accept global auto accept upgrade
        * Do not accept user auto accept upgrade
            * Downloads go to user download dir
        * Accept user auto accept upgrade
            * Downloads go to new global download dir / hashed id

* Global set and User not set
    * Do not accept global upgrade
        * User upgrade not prompted
            * Files downloaded to global download dir
                * This interation is a little undesirable but will be cleaned up in #5376
        * Enable autoaccept for user
            * Files downloaded to global download dir / hashed id
                * This interation is a little undesirable but will be cleaned up in #5376
    * Accept global upgrade
        * User upgrade not prompted
            * Files downloaded to global download dir
                * This interation is a little undesirable but will be cleaned up in #5376
        * Enable autoaccept for user
            * Files downloaded to global download dir / hashed id
                * This interation is a little undesirable but will be cleaned up in #5376

* New Program Run (no qtox.ini)
    * No prompts
    * Default global autoaccept dir set to new global download dir
    * Same interactions as above

* Open directory in friend about page now opens user download dir
* Reset resets friend download dir to global dir / hashed id
* Download dir follows global dir change if default/reset

Tests can be run by opening a user profile after downgrading qtox, this should reset behavior so you can experience the upgrade logic again :) (with the side affect of disabling autoaccept for all users)


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5380)
